src/gpuarray_array_collectives.c

```diff
@@ -0,0 +1,108 @@
+#include "gpuarray/array.h"
+#include "gpuarray/buffer_collectives.h"
+#include "gpuarray/collectives.h"
+#include "gpuarray/error.h"
+
+#include "private.h"
+
+static inline int find_total_elems(const GpuArray* array)
+{
```

Comment:
You can always add a check inside the nccl wrappers to make sure that we don't actually bust 4G.